gslib/utils/copy_helper.py

```diff
@@ -670,59 +673,67 @@ def ConstructDstUrl(src_url,
     # exp_src_url_tail should become /123
     # Note: mv.py code disallows wildcard specification of source URL.
-    recursive_move_to_new_subdir = True
     exp_src_url_tail = (exp_src_url.url_string[len(src_url.url_string):])
     dst_key_name = '%s/%s' % (exp_dst_url.object_name.rstrip('/'),
                               exp_src_url_tail.strip('/'))
 
-  elif src_url_names_container and (exp_dst_url.IsCloudUrl() or
-                                    exp_dst_url.IsDirectory()):
-    # Case 2.  Container copy to a destination other than a file.
-    # Build dst_key_name from subpath of exp_src_url past
-    # where src_url ends. For example, for src_url=gs://bucket/ and
-    # exp_src_url=gs://bucket/src_subdir/obj, dst_key_name should be
-    # src_subdir/obj.
-    src_url_path_sans_final_dir = GetPathBeforeFinalDir(src_url, exp_src_url)
-
-    dst_key_name = exp_src_url.versionless_url_string[
-        len(src_url_path_sans_final_dir):].lstrip(src_url.delim)
-    # Handle case where dst_url is a non-existent subdir.
-    if not have_existing_dest_subdir:
-      dst_key_name = dst_key_name.partition(src_url.delim)[-1]
-    # Handle special case where src_url was a directory named with '.' or
-    # './', so that running a command like:
-    #   gsutil cp -r . gs://dest
-    # will produce obj names of the form gs://dest/abc instead of
-    # gs://dest/./abc.
-    if dst_key_name.startswith('.%s' % os.sep):
-      dst_key_name = dst_key_name[2:]
-
   else:
-    # Case 3.
-    dst_key_name = exp_src_url.object_name.rpartition(src_url.delim)[-1]
-
-  if (not recursive_move_to_new_subdir and
-      (exp_dst_url.IsFileUrl() or _ShouldTreatDstUrlAsBucketSubDir(
-          have_multiple_srcs, exp_dst_url, have_existing_dest_subdir,
-          src_url_names_container, recursion_requested))):
-    if exp_dst_url.object_name and exp_dst_url.object_name.endswith(
-        exp_dst_url.delim):
-      dst_key_name = '%s%s%s' % (exp_dst_url.object_name.rstrip(
-          exp_dst_url.delim), exp_dst_url.delim, dst_key_name)
+    if preserve_src_top_level_dirs or (src_url_names_container and
+                                       (exp_dst_url.IsCloudUrl() or
+                                        exp_dst_url.IsDirectory())):
+      # Case 2.  Container copy to a destination other than a file.
+      # Build dst_key_name from subpath of exp_src_url past
+      # where src_url ends. For example, for src_url=gs://bucket/ and
+      # exp_src_url=gs://bucket/src_subdir/obj, dst_key_name should be
+      # src_subdir/obj.
+      src_url_path_sans_final_dir = GetPathBeforeFinalDir(src_url, exp_src_url)
+      dst_key_name = exp_src_url.versionless_url_string[
+          len(src_url_path_sans_final_dir):].lstrip(src_url.delim)
+
+      # Handle case where dst_url is a non-existent subdir unless
+      # multiple top-level sources, in which case preserve source hierarchy.
+      if not preserve_src_top_level_dirs:
```

Comment:
If we remove the mv condition from the above if condition, cand we drop this condition in that case. From the way I understand it now, the behavior for mv and cp is going to be the same. In that case, I think if we drop the perform_mv condition, do we still need this check?

Reply:
The `mv` code in the "Case 1" block is:
```
if (global_copy_helper_opts.perform_mv and recursion_requested
    and src_url_names_container and not preserve_src_top_level_dirs):
  exp_src_url_tail = (exp_src_url.url_string[len(src_url.url_string):])
  dst_key_name = '%s/%s' % (exp_dst_url.object_name.rstrip('/'),
                              exp_src_url_tail.strip('/'))
```
To get to this block you're commenting on, you have to meet this condition (combining "Case 2" and `not preserve_src_top_level_dirs`):
```
if src_url_names_container and (exp_dst_url.IsCloudUrl() or exp_dst_url.IsDirectory():
  src_url_path_sans_final_dir = GetPathBeforeFinalDir(src_url, exp_src_url)
  dst_key_name = exp_src_url.versionless_url_string[
          len(src_url_path_sans_final_dir):].lstrip(src_url.delim)
  dst_key_name = dst_key_name.partition(src_url.delim)[-1]
```

It looks like there isn't an overlap of logic we can condense.